runtime/pubsub/local.go

```diff
@@ -0,0 +1,163 @@
+//go:build encore_local
+
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/nsqio/go-nsq"
+
+	"encore.dev/pubsub/utils"
+	"encore.dev/rlog"
+	"encore.dev/runtime/config"
+)
+
+// localTopic is the nsq implementation of pubsub.Topic. It exposes methods to publish
+// and subscribe to messages of a topic
+type localTopic[T any] struct {
+	name      string
+	cfg       *TopicConfig
+	addr      string
+	producer  *nsq.Producer
+	consumers map[string]*nsq.Consumer
+	idSeq     uint32
+}
+
+// message is a local representation of a topic published to NSQ.
+// it wraps the raw data with an ID and an Attribute map
+type message[T any] struct {
+	ID         string
+	Attributes map[string]string
+	Data       T
+}
+
+// handler implements the interface of an nsq.Hander. It is used to Unmarshal a message
+// and forward it to a subscriber.
+type handler[T any] struct {
+	sub Subscriber[T]
+	cfg *SubscriptionConfig
+}
+
+// HandleMessage unmarshals a message into the message type and forwards it to the subscriber
+func (h *handler[T]) HandleMessage(m *nsq.Message) (err error) {
+	defer func() {
+		if err != nil {
+			rlog.Error("failed to handle message", "messageId", m.ID, "err", err)
+		}
+		if !m.HasResponded() {
+			policy := h.cfg.RetryPolicy
+			retry, delay := utils.GetDelay(policy.MaxRetries, policy.MinRetryDelay, policy.MaxRetryDelay, m.Attempts)
+			if !retry {
+				rlog.Info("depleted message retries. Dropping message", "messageId", m.ID)
+				m.Finish()
+				return
+			}
+			m.RequeueWithoutBackoff(delay)
+		}
+	}()
+	// create a message to unmarshal the raw nsq body into
+	msg := &message[T]{}
+	err = json.Unmarshal(m.Body, msg)
+	if err != nil {
+		return err
+	}
+	// unmarshal the attributes and write them to the message type
+	err = utils.UnmarshalFields(msg.Attributes, &msg.Data, "pubsub-attr")
```

Comment:
Should we unmarshal the attributes into the message?

When we marshal the original message we don't remove them - so those attributes are already inside the `msg.Data` object at this point as `json.Unmarshal` has done this already.

So I guess my question here is - are we either:
1. Missing the code to _remove the attribute fields_ during the original marshalling?
2. Doing a redundant extra unmarshal here?

Reply:
Hmm, yeah. I discussed this with Andre, and we concluded that we don't want to constrain the user too much. The user can choose to exclude the field from the body with `json:"-"`, but the default would be to allow either or. You're correct that we will do a redundant unmarshal, but at this point I wanted to avoid the complexity of implementing custom json serialization. 

I think the next step is indeed to add generated code, but wanted to avoid colliding with andre's changes